x/ibc/23-commitment/types/merkle.go

```diff
@@ -154,31 +183,292 @@ func (MerkleProof) GetCommitmentType() exported.Type {
 }
 
 // VerifyMembership verifies the membership pf a merkle proof against the given root, path, and value.
-func (proof MerkleProof) VerifyMembership(root exported.Root, path exported.Path, value []byte) error {
+func (proof MerkleProof) VerifyMembership(specs []*ics23.ProofSpec, root exported.Root, path exported.Path, value []byte) error {
 	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
-		return errors.New("empty params or proof")
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
 	}
 
-	runtime := rootmulti.DefaultProofRuntime()
-	return runtime.VerifyValue(proof.Proof, root.GetHash(), path.String(), value)
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return sdkerrors.Wrap(ErrInvalidProof, "path is not a merkle path for a merkle proof")
+	}
+
+	if len(proof.Proofs) != len(mpath.KeyPaths) || len(proof.Proofs) != len(specs) {
+		return sdkerrors.Wrapf(ErrInvalidProof, "invalid chained proof. chained proof length %d, spec length %d, path length %d must all be equal",
+			len(proof.Proofs), len(specs), len(mpath.KeyPaths))
+	}
+
+	for i, p := range proof.Proofs {
+		// While the proofs go from lowest subtree to the final tree, the path is defined from the root
+		// down to the leaf. Thus, we must pass in subpaths in reverse order during chained proof verification
+		subpath := []byte(mpath.KeyPaths[len(mpath.KeyPaths)-i-1].String())
+		existProof, ok := p.Proof.(*ics23.CommitmentProof_Exist)
+		if !ok {
+			return sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
+		}
+		// For subtree verification, we simply calculate the root from the proof and use it to prove
+		// against the value
+		subroot, err := existProof.Exist.Calculate()
+		if err != nil {
+			return sdkerrors.Wrap(ErrInvalidProof, err.Error())
+		}
+		if i != len(proof.Proofs)-1 {
```

Comment:
Looking at this code, I think it would make more sense going from top-to-bottom, so we don't have to pre-calculate the expected root each time.

Or is the embedded key-value not passed along with the proof.

I mean, I first check there is a valid proof from `header.AppHash` to `("ibc", <sub-hash>)`.
If true, then I check if there is a valid proof from `<sub-hash>` to `(<packet-key>, <packet-value>)`.

I assume that `<sub-hash>`, `<packet-key>` and `<packet-value>` are sent along with the message. If not, then my inversion doesn't work. Can you point me to all info we have in the calling function (or is this interface fixed). This function does take root hash, all keys and packet-value, so I do see why you have to go reverse... just seems a bit odd

Reply:
subhash would not be sent with the message